### PR TITLE
added recursive getGradeLogs function to handle bookmarked pagination

### PR DESCRIPTION
**Pull request by mbogor**
#### What's this PR do?
fixes pagination on the platformGraduation script
#### Where should the reviewer start?
jobs/training/pollCanvas
#### How should this be manually tested?
run the file and check the graduates slack channel
#### What are the relevant tickets?
https://admin.crisistextline.org/jira/browse/INT-452

### CODE REVIEW

File: canvas.js [hunk 1]
```diff
@@ -24,6 +25,56 @@ function createOptions(reqOptions){
...
+
+		function newPage(nextPage){
+			if(nextPage) options.url = `${nextPage}`;
+			return fetch(options)
+			.then(function(response){
+				var link = linkParser(response.headers.link);
+				if(link.last) {
+					if(!payload) payload = response.body;
+					else payload.events = payload.events.concat(response.body.events);
+					console.log('PAYLOAD', payload.events.length)
```
Comment (apackin) [inline canvas.js]: Keeping this?

File: canvas.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
 'use strict';
 const canvas = {};
 const fetch = require('requestretry');
+const linkParser = require('parse-link-header');
```
Comment (apackin) [inline canvas.js]: Cool package

File: canvas.js [hunk 3]
```diff
@@ -24,6 +25,56 @@ function createOptions(reqOptions){
...
+			});
+		}
+		return newPage();
+};
+
+//get quiz submissions
+canvas.getQuizSubmissions = function(courseID, quizID){
+	const url = `https://crisistextline.instructure.com/api/v1/courses/${courseID}/quizzes/${quizID}/submissions?per_page=100`;
+	const options = createOptions({
+		url: url
```
Comment (apackin) [inline canvas.js]: Can do {url}

File: jobs/training/pollCanvasForGraduatedUsersThenCreatePlatformAccount.js [hunk 4]
```diff
@@ -1,18 +1,20 @@
...
-var CronJob = require('cron').CronJob;
-var SLACK_CHANNEL = '#graduates';
+const canvas = require('../../canvas.js');
+const fetch = require('node-fetch');
+const crypto = require('crypto');
+const internalRequest = require('request');
+const moment = require('moment');
+const throttler = require('throttled-request')(internalRequest);
+const CronJob = require('cron').CronJob;
+const SLACK_CHANNEL = '#graduates';
```
Comment (apackin) [inline jobs/training/pollCanvasForGraduatedUsersThenCreatePlatformAccount.js]: 😍

File: test/recurOpenShifts.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ describe('recurOpenShifts', function() {
   describe('returnMaxOpenShiftCountForTime', function() {
 
     it('should return max open shift count for a time', function () {
-      assert.equal(recurOpenShifts.returnMaxOpenShiftCountForTime(targetTime), 19);
+      assert.equal(recurOpenShifts.returnMaxOpenShiftCountForTime(targetTime), 29);
```
Comment (apackin) [inline test/recurOpenShifts.js]: Thank you

File: canvas.js [hunk 6]
```diff
@@ -24,6 +25,56 @@ function createOptions(reqOptions){
...
+			retryDelay: 1000
+		});
+		var payload,
+				nextPage;
+
+		function newPage(nextPage){
+			if(nextPage) options.url = `${nextPage}`;
+			return fetch(options)
+			.then(function(response){
+				var link = linkParser(response.headers.link);
```
Comment (apackin) [inline canvas.js]: 'var' is like so last year 😝

**Comment by apackin** [comment]
Seems to be working over all. I'm getting ```[Wed, Nov 2, 2016 1:38 PM -04:00] Canvas call to get grade logs failed: TypeError: Cannot read property 'last' of null
[Wed, Nov 2, 2016 1:38 PM -04:00] Call to Canvas to get grade logs for course 73 has failed, error:  {}
[Wed, Nov 2, 2016 1:38 PM -04:00] Canvas call to get grade logs failed: TypeError: Cannot read property 'last' of null
[Wed, Nov 2, 2016 1:38 PM -04:00] Call to Canvas to get grade logs for course 64 has failed, error:  {}
[Wed, Nov 2, 2016 1:38 PM -04:00] Canvas call to get grade logs failed: TypeError: Cannot read property 'last' of null
[Wed, Nov 2, 2016 1:38 PM -04:00] Call to Canvas to get grade logs for course 72 has failed, error:  {}
[Wed, Nov 2, 2016 1:38 PM -04:00] Canvas call to get grade logs failed: TypeError: Cannot read property 'last' of null
[Wed, Nov 2, 2016 1:38 PM -04:00] Call to Canvas to get grade logs for course 58 has failed, error:  {}
[Wed, Nov 2, 2016 1:38 PM -04:00] Canvas call to get grade logs failed: TypeError: Cannot read property 'last' of null
[Wed, Nov 2, 2016 1:38 PM -04:00] Call to Canvas to get grade logs for course 93 has failed, error:  {}``` for about 20 courses.

**Comment by mbogor** [comment]
i added error handling for 500 errors from canvas (a new one!) and the errors assaf was getting went away.

File: canvas.js [hunk 7]
```diff
@@ -24,6 +25,84 @@ function createOptions(reqOptions){
 	return reqOptions;
 
 }
+//finds all grade changes
+canvas.getGradeLogs = function(courseID){
```
Comment (tongxiang) [inline canvas.js]: this function is hard to read. Would you mind adding some documentation, particularly within the promise chain? 

Thanks!

File: canvas.js [hunk 8]
```diff
@@ -24,6 +25,84 @@ function createOptions(reqOptions){
...
+		const url = `https://crisistextline.instructure.com/api/v1/audit/grade_change/courses/${courseID}?per_page=100`;
+		const options = createOptions({
+			url,
+			retryDelay: 1000
+		});
+		let payload;
+
+		function newPage(nextPage){
```
Comment (tongxiang) [inline canvas.js]: would be great to have some documentation here

File: canvas.js [hunk 9]
```diff
@@ -24,6 +25,84 @@ function createOptions(reqOptions){
...
+						payload.linked.users = payload.linked.users.concat(response.body.linked.users);
+					}
+					return newPage(nextPage); 
+				}
+			})
+			.catch(function(err){
+				CONSOLE_WITH_TIME(`Canvas call to get grade logs failed: ${err}`);
+			});
+		}
+		return newPage();
```
Comment (tongxiang) [inline canvas.js]: Clever recursive solution!

**Comment by tongxiang** [comment]
ONe comment on documentation, otherwise looks good. 👍

File: canvas.js (hunk 8 above)
Comment (mbogor) [inline canvas.js]: done!

**Pull request merged**